File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> spiral = new ArrayList<>();
        
        int left = 0;
        int right = matrix[0].length - 1;
        int top = 0;
        int bottom = matrix.length - 1;

        while (top <= bottom && left <= right) {
            // Traverse from left to right along the top row
            for (int i = left; i <= right; i++) {
                spiral.add(matrix[top][i]);
            }
            top++;

            // Traverse from top to bottom along the right column
            for (int i = top; i <= bottom; i++) {
                spiral.add(matrix[i][right]);
            }
            right--;

            // Traverse from right to left along the bottom row
            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    spiral.add(matrix[bottom][i]);
                }
                bottom--;
            }

            // Traverse from bottom to top along the left column
            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    spiral.add(matrix[i][left]);
                }
                left++;
            }
        }

        return spiral;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rows = len(matrix)
        cols = len(matrix[0])
        left = 0
        right = cols - 1
        bottom = rows - 1
        top = 0
        # [1,2,3,4,8,12,11,10,9,5,6,7,6]
        # [1,2,3,4,8,12,11,10,9,5,6,7]

        answer = []
        # rows != cols
        # 3 * 4
        # l = 0 , r = 3, b = 2, t = 0
        # t=0 , 0 -> 4
        # top = 1, [1,2,3,4]
        # right = 3, 1 -> 3
        # right = 2, [1,2,3,4,8,12]
        # bottom = 2, 2 -> -1
        # bottom = 1, [1,2,3,4,8,12,11,10,9]
        # left = 0, 1 -> 0
        # left = 1 , [1,2,3,4,8,12,11,10,9, 5]
        
        #t = 1, 1 -> 3
        # top = 2, [1,2,3,4,8,12,11,10,9, 5, 6, 7]
        
        # right = 2, 2 -> 1
        # no loop
        # right = 1

        # right -> left
        # bottom = 1, 1 -> 0
        # revisiting 6

        # 
        while left <= right and top <= bottom:
            # on top
            # left -> right
            for i in range(left, right + 1):
                answer.append(matrix[top][i])
            top += 1

            # on right
            # top -> bottom
            # print(left, right, bottom, top)
            # 0 2 3 1
            for i in range(top, bottom + 1):
                answer.append(matrix[i][right])
            right -= 1

            # on bottom
            # right -> left
            if top <= bottom:
                for i in range(right , left - 1, -1):
                    answer.append(matrix[bottom][i])
                bottom -= 1
            # on the left
            # b -> t
            if left <= right:
                for i in range(bottom, top - 1, -1):
                    answer.append(matrix[i][left])
                left += 1
        
        return answer